### ISSUE TITLE
check-cloudtrail filter not working as expected?

### ISSUE THREAD

**Issue by lithiumoxide** [cloud-custodian/cloud-custodian#6134]
### Summary
I'm looking for some clarification on how the `check-cloudtrail` filter works within `resource: account`. I'm trying to set up policies to find cases where CloudTrail is or is not running in a given account, but I can't figure out a filter configuration to match the account for either case.

### Setup

I have an account with CloudTrail running: there are two trails and each one is a multi-region trail.

Right now I have this filter in my policy:

```
    filters:
      - type: check-cloudtrail
        multi-region: true
        running: true
```

### Result

My expectation here is that Custodian would find that the account resource matches this filter, as the trails are both running _and_ multi-region. However, when I run the Lambda function for the policy, I see the following result:

```
Filtered from 1 to 0 account
```

If I also removed `multi-region: true` and have only `running: true` in the filter I get the same result: the account does not match the filter. Similarly, I've also tried the `global-events` option.

In short, it seems that I can't figure out a configuration of the filter to match the account.

### End goal

My goal in this task is to set up a policy to find instances when CloudTrail is _not_ enabled in the account. All my accounts have CloudTrail enabled, so my method here is to set up a policy to find CloudTrail-enabled accounts and when that policy works, I will set the filters to their inverse, i.e. `running: false`, etc.

**Comment by kapilt** [comment]
cloudtrail filter on account is setup for an inverse match, ie find the case where those things don't exist.

given the cardinality increases of cloudtrail, we ended up adding in a separate cloudtrail resource which allows might be more appropriate for your use case, and then using it as an embedded policy with the missing filter on the account to detect non existence.

**Comment by lithiumoxide** [comment]
Hi @kapilt. Thanks very much for that information. You've pretty much described exactly what I want to do: use `account` to find cases of non-existent trails, and `cloudtrail` to find misconfigured trails. I've already deployed the `cloudtrail` policy and it works well.

I'm experimenting with the `missing` filter in `account` now and checked the schema, but I'm not sure if I'm using it correctly. To test it, I removed `Trail1` from the account; only `Trail2` exists now. Here is my policy name and filter:

```
policies:
  - name: audit-cloudtrail-enable
  	...
  	...
    filters:
      - or:
        - type: missing
          policy:
            resource: cloudtrail
            name: 'Trail1'
        - type: missing
          policy:
            resource: cloudtrail
            name: 'Trail2'
```

My intention here is to have the policy match a case where either of my two trails are missing, and then notify me (I have notifications set up as part of the policy actions). I removed `Trail1` from the account, but with the above filter I'm not getting any notification.

So, I inspected the Lambda function and found the following. In `config.json` I find the following JSON. Note how the `name` attribute within the `policy` is set to the name of the c7n policy, and not the name I defined in the filter - this leads me to believe I have it set up incorrectly.

```
      "filters": [
        {
          "or": [
            {
              "type": "missing",
              "policy": {
                "resource": "cloudtrail",
                "name": "audit-cloudtrail-enable"
              }
            },
            {
              "type": "missing",
              "policy": {
                "resource": "cloudtrail",
                "name": "audit-cloudtrail-enable"
              }
            }
          ]
        }
      ]
```

In the execution logs I see the following two lines:

```
Filtered from 1 to 1 cloudtrail
Filtered from 1 to 0 account
```

My question now is: Am I building the filter correctly? Or have I misunderstood how it works?

Thanks again for all your help with this.

**Comment by lithiumoxide** [comment]
Just coming back to this: I did indeed misunderstand how it worked, but I'm happy to say that I found what I was doing wrong. I think I missed your point about it being an _embedded_ policy, so when I re-read your comments I took a different approach.

For future reference and for anyone else who comes across this, I want to show my example of how it works:

```
policies:
  - name: audit-cloudtrail-enable
    description: This policy evaluates whether CloudTrail is enabled.
    resource: account
    ...
    ...
    filters:
      - type: missing
        policy:
          resource: cloudtrail
          filters:
          - or:
            - type: value
              key: Name
              value: MyCloudTrailA
            - type: value
              key: Name
              value: MyCloudTrailB
    ...
    ...
```

It's really quite simple in retrospect.

Thanks again for your help!
